### ISSUE TITLE
ENH: add groupby(..).sort_values()

### ISSUE THREAD

**Issue by jreback** [pandas-dev/pandas#15947]
might be useful to expose this as a top-level groupby method for intra-group sorting. Anyone have a good usecase for this?
```
In [10]: df = DataFrame({'A': [1, 1, 2, 2, 2], 'B': [2, 1, 3, 3, 1], 'C':[1, 1, 2, 1, 1]})

In [11]: df
Out[11]: 
   A  B  C
0  1  2  1
1  1  1  1
2  2  3  2
3  2  3  1
4  2  1  1

In [12]: df.groupby('A').apply(lambda x: x.sort_values(['B', 'C']))
Out[12]: 
     A  B  C
A           
1 1  1  1  1
  0  1  2  1
2 4  2  1  1
  3  2  3  1
  2  2  3  2

```

**Comment by jorisvandenbossche** [comment]
What would be the difference with just sorting it by those three columns? `df.sort_values(['A', 'B', 'C'])` ?

(one possible difference I see is that you could retain the original order in `'A'` if you passed `sort=False` to the groupby, but not sure that is worth it)

**Comment by jreback** [comment]
@jorisvandenbossche NO, that's the point, it would basically be a convenience method to avoid exactly this problem. alternatively maybe we could just warn/raise.

**Comment by jorisvandenbossche** [comment]
What do you mean with warn/raise? What is there to warn for?

**Comment by jreback** [comment]
I mean sorting after grouping is non-performant, you should always sort *before*. we certainly *can* fix it here (by providing a convenience method). If we choose *not* to fix it directly, then I think showing a ``PerformanceWarning`` is appropriate (though maybe just fix it).

**Comment by randomgambit** [comment]
@jreback @jorisvandenbossche its funny because I was thinking about this problem this morning. 

Here is a very common set up. A large dataset contains news (identified by a `story_id`) and for the same news you have several entities (identified by an `entity_id`): IBM, APPLE, etc.

What you wanna do is get the most **relevant** entity for each news. That is:

`df.groupby('story_id').apply(lambda x:  x.sort_values(by = 'relevance', ascending = False))`

followed by a `drop_duplicates(subset ='story_id', keep= 'first')`

very common use case

**Comment by randomgambit** [comment]
I confirm this is veeeeery slow :(

**Comment by jorisvandenbossche** [comment]
And again, why don't you use `df.sort_values(['story_id', 'relevance'])` ?

**Comment by randomgambit** [comment]
yes, I did. but I think @jreback 's point is that, while conceptually equivalent, the `groupby` solution is INSANELY slow :D

**Comment by jorisvandenbossche** [comment]
OK, but my point originally was: if you can simply use `sort_values`, is there a *need* to add it to `groupby` ? (and thus when adding, doing effort to make sure it is not insanely slow)

**Comment by randomgambit** [comment]
in my opinion, nope. just throw some performance warning like : "hey dummy what are you doing??? just use sort_values!!!! "

;-)

**Comment by jacquespeeters** [comment]
Because the sum of all sorting' complexity inside groups is lower than the sorting' complexity on the whole dataset. `groupby` complexity doesn't change.

There is a huge performance difference when dataset is big, and groups small. It is right now slower, not because of the complexity, but because of `apply()`.

A good way to compare both operations complexity would be : 
```
df.groupby('story_id').apply(lambda x: x.sort_values(by = 'relevance', ascending = False))
df.apply(lambda x: x.sort_values(by = 'relevance', ascending = False)).groupby('story_id')
```
